src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+	}
+	
+	static class RegexList extends RegexElement {
+		private List<RegexElement> elements = new ArrayList<>();
+		
+		public void addElement(RegexElement element) {
+			elements.add(element);
+		}
+		
+		public List<RegexElement> getElements() {
+			return elements;
+		}
+		
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder();
+			for (RegexElement element:elements) {
+				sb.append(element.toString());
+			}
+			return sb.toString();
+		}
+	}
+	
+	/**
+	 * Group with the optional modifier
+	 */
+	static class OptionalRegexGroup extends RegexList {
+		
+		@Override
+		public String toString() {
+			return "(" + super.toString() + ")?";
+		}
+	}
+	
+	static class RegexToken extends RegexElement {
+		private String token;
+		
+		public RegexToken(String token) {
+			this.token = token.trim();
+		}
+		
+		public String getToken() {
+			return token;
+		}
+		
+		@Override
+		public String toString() {
+			return Pattern.quote(
+					LicenseCompareHelper.NORMALIZE_TOKENS.getOrDefault(token.toLowerCase(), token)) + 
+					"\\s*";
+		}
+	}
+	
+	static class RegexPattern extends RegexElement {
+		private String pattern;
+		
+		public RegexPattern(String pattern) {
+			this.pattern = pattern;
+		}
+		
+		public String getPattern() {
+			return pattern;
+		}
+		
+		@Override
+		public String toString() {
+			return "(" + pattern + ")"; // We always treat the pattern as a group
+		}
+
+		/**
+		 * @param pattern pattern to set
+		 */
+		public void setPattern(String pattern) {
+			this.pattern = pattern;
+		}
+	}
+
+	private String template;
+	
+	/**
+	 * Top level regex
+	 */
+	private RegexList regexPatternList = new RegexList();
+	
+	
+	private int optionalNestLevel = 0;
+	
+	private List<OptionalRegexGroup> optionalGroups = new ArrayList<>();
+
+	/**
+	 * Generates regular expressions from a license or exception template
+	 * @throws SpdxCompareException 
+	 */
+	public TemplateRegexGenerator(String template) throws SpdxCompareException {
+		this.template = template;
+		parseTemplate();
+	}
+	
+	/**
+	 * Parses the template generating the regular expression
+	 * @throws SpdxCompareException 
+	 */
+	private void parseTemplate() throws SpdxCompareException {
+		try {
+			SpdxLicenseTemplateHelper.parseTemplate(LicenseCompareHelper.removeCommentChars(template), this);
+		} catch (LicenseTemplateRuleException e) {
+			throw new SpdxCompareException("Invalid template rule found during filter: "+e.getMessage(),e);
+		} catch (LicenseParserException e) {
+			throw new SpdxCompareException("Invalid template found during filter: "+e.getMessage(),e);
+		}
+	}
+	
+	/**
+	 * @return the complete regular expression for the template
+	 */
+	public String getCompleteRegex() {
+		return regexPatternList.toString();
+	}
+	
+	/**
+	 * @param wordLimit number of non optional words to include in the pattern
+	 * @return a regex to match the start of the license per the template
+	 */
+	public String getStartRegex(int wordLimit) {
```

Comment:
Might this be private?  Or are there cases where callers (e.g. `LicenseCompareHelper`) actually need the start regex independently of the entire thing?

Reply:
Agree - I'll make these private